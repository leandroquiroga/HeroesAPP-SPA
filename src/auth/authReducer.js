import { types } from "../types/type";



export const authReducer = (state = {}, action) => {
  const { type, payload } = action;
  
  switch (type) {
    case types.login:
      return {
        ...payload,
        logged: true,
      };
    case types.logout:
      return {
        logged: false,
      };
    
    default:
      return state;
  }
}