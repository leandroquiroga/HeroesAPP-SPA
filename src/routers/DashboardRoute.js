import { Fragment } from "react";
import { Route, Routes } from "react-router-dom";

import { MarvelScreen } from '../components/Category/MarvelScreen';
// import { HerosApp } from '../HerosApp';
import { DCScreen } from './../components/Category/DCScreen';
import { SearchScreen } from './../components/Search/Search';
import { NavbarScreen } from './../components/UI/Navbar';
import { Heroes } from './../components/Heores/Heroes';
import { Main } from "../components/UI/Main";

export const DashboardRoute = () => {
  return (
    <Fragment>
      <NavbarScreen />
      <Routes>
        <Route path="/" element={<Main />} />
        <Route path="marvel" element={<MarvelScreen />} />
        <Route path="dc" element={<DCScreen />} />
        <Route path="hero/:heroID" element={<Heroes />} />
        <Route path="search" element={<SearchScreen />} />
      </Routes>
    </Fragment>
  );
};
