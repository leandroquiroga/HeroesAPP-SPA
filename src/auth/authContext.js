import { createContext, useEffect, useReducer, useState } from "react";
import { signInWithEmailAndPassword, signOut ,createUserWithEmailAndPassword } from 'firebase/auth';
import { authReducer } from "./authReducer";
import { auth } from './../firebase';
import { heroReducer } from "../helpers";

// Retorna el usuario si lo encuentra en localStorage si no retorna el logged en false
const init = () => JSON.parse(localStorage.getItem('user')) || {logged: false}

// Retorna los hero si lo encuntra en localStorage si no retorna un array 
const initHero = () => JSON.parse(localStorage.getItem('hero')) || [];

export const AuthContext = createContext();

export const AuthProvaider = (props) => {
  const [user, dispatch] = useReducer(authReducer, {}, init);
  const [currentUser, setCurrentUser] = useState({});
  const [hero, dispatchHero] = useReducer(heroReducer, [], initHero);

  // Cuando inicia la aplicacion va a estar observando si hay un cambio
  // de usuario.
  useEffect(() => {
    auth.onAuthStateChanged((user) => setCurrentUser(user))
  }, []);

  // Esta pendiente al user para ir almancendolo en el localStorage
  useEffect(() => {
    localStorage.setItem('user', JSON.stringify(user))
    return () => {
      if (!user) return;
    }
  }, [user]);

  // Esta pendiente al hero, cuando detecta un cambio lo almacena en el localStorage
  useEffect(() => {
    localStorage.setItem('hero', JSON.stringify(hero));
    return () => {
      if (!hero) return;
    }
  }, [hero])
  


  const singUp = (email, password) => createUserWithEmailAndPassword(auth, email, password);

  const logIn = (email, password) => signInWithEmailAndPassword(auth, email, password);
  
  const logOut = (auth) => signOut(auth);

  const value = {
    user,
    dispatch,
    singUp,
    logIn,
    logOut,
    currentUser,
    dispatchHero,
    hero,
  }
  return (
    <AuthContext.Provider value={value}>
      {props.children}
    </AuthContext.Provider>
  )
}