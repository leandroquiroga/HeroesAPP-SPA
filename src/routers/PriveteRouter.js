

import React, { useContext, useEffect } from 'react';
import { AuthContext } from '../auth/authContext';
import { Navigate, useLocation } from 'react-router-dom';

export const PriveteRouter = ({children}) => {
  
  const { currentUser } = useContext(AuthContext);
  const { pathname, search } = useLocation();
  
  // Esta atento al cambio del nuesto pathname del useLocation,
  // cuando detecte un cambio checkea si exite el search, en caso de que 
  // sea correcto se alamacena en el localStorage, sino se almancena el pathname
  useEffect(() => {
    localStorage.setItem('lastPath', pathname + (search && search) )
  },[pathname, search])
  
  return (currentUser) ?
          children : <Navigate to='/login'/> 
};


// En caso que el usuario este autenticado
// le permite acceso a la aplicacion 
// en caso contrario lo redirige 
// al que dse autentifique