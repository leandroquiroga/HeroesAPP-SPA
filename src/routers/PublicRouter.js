import { useContext } from "react"
import { Navigate } from "react-router-dom";
import { AuthContext } from "../auth/authContext";


export const PublicRouter = ({ children }) => {
  
  const { user } = useContext(AuthContext);
  
  // Si el usuario no esta autotenticado
  // muestra el el componente de login
  // en caso contrario muestra la pagina de inicio
  return (!user.logged) ?
        children
        :<Navigate to='/' />
};