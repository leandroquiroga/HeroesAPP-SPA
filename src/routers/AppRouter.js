import React from 'react';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import { PriveteRouter } from './PriveteRouter';
import { PublicRouter } from './PublicRouter';
import { LoginScreen } from './../components/Login/LoginScreen';
import { DashboardRoute } from './DashboardRoute';
import { RegisterScreen } from '../components/Register/RegisterScreen';

export const AppRouter = () => {
  return (
    <BrowserRouter>
      <Routes>
        
        <Route path="/login" element={
          <PublicRouter >
            <LoginScreen />
          </PublicRouter>
        } />
        
        <Route path='/register' element={ 
          <PublicRouter>
            <RegisterScreen />
          </PublicRouter>
        } />

        <Route path="/*" element={
          <PriveteRouter>
              <DashboardRoute />
          </PriveteRouter>
          } />
      </Routes>
    </BrowserRouter>
  );
};
