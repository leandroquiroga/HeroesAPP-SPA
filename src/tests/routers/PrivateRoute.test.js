import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { AuthContext } from '../../auth/authContext';
import { DashboardRoute } from '../../routers/DashboardRoute';
import { PriveteRouter } from './../../routers/PriveteRouter';


jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  Navigate: () => <span>Saliendo...</span>
}))

describe('Testing on <PrivateRouter />', () => {
  
  Storage.prototype.setItem = jest.fn();
  
  test('This test should show corretly the private comoponet if user is authentucated', () => {

    const fakeAuth = {
      currentUser: true,
    };

    const wapper = mount(
      <AuthContext.Provider value={fakeAuth}>
        <MemoryRouter initalEntries={['/']}>
          <PriveteRouter>
            <h1>Componente Privado</h1>
          </PriveteRouter>
        </MemoryRouter>
      </AuthContext.Provider>
    );
    
    expect(wapper.text().trim()).toBe('Componente Privado')
  });


  test('This test should show the component <DashboardRoute /> without any hero', () => {
    
    const textInfo = 'No hay superheroes en la lista de favoritos...';

    const contextValue = {
      currentUser: true,
      user: {
        name: 'Leandro'
      },
      hero: []
    };

    const wapper = mount(
      <AuthContext.Provider value={contextValue}>
        <MemoryRouter initalEntries={['/']}>
          <PriveteRouter>
            <DashboardRoute />
          </PriveteRouter>
        </MemoryRouter>
      </AuthContext.Provider>
    );

    expect(wapper).toMatchSnapshot();
    expect(wapper.find('p').text().trim()).toBe(textInfo);
    expect(contextValue.hero.length).toBeLessThan(1);
  });


  test('This test should show the component <DashboardRoute /> with at least one hero', () => {
    
    const contextValue = {
      currentUser: true,
      user: {
        name: 'Leandro'
      },
      hero: [
        {
          'id': 'dc-batman',
          'superhero': 'Batman',
          'publisher': 'DC Comics',
          'alter_ego': 'Bruce Wayne',
          'first_appearance': 'Detective Comics #27',
          'characters': 'Bruce Wayne'
        },
      ],
    };

    const wapper = mount(
      <AuthContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={['/']}>
          <PriveteRouter>
            <DashboardRoute />
          </PriveteRouter>
        </MemoryRouter>
      </AuthContext.Provider>
    );
    
    expect(wapper).toMatchSnapshot();
    expect(wapper.find('h5').text().trim()).toBe('Batman');
    expect(contextValue.hero.length).toBeGreaterThanOrEqual(1);
  });

  test('This test should not show the private component if the user is not authenticated', () => {

    const contextValue = {
      currentUser: false,
    };

    const wapper = mount(
      <AuthContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={['/']}>
          <PriveteRouter>
            <h1>Componente Privado</h1>
          </PriveteRouter>
        </MemoryRouter>
      </AuthContext.Provider>
    );

    expect(wapper.text().trim()).toBe('Saliendo...');
  });
});