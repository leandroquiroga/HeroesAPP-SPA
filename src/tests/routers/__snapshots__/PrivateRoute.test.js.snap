// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing on <PrivateRouter /> This test should show the component <DashboardRoute /> with at least one hero 1`] = `
Array [
  <nav
    className="navbar navbar-expand-lg navbar-dark bg-dark px-2"
  >
    <section
      className="container-fluid"
    >
      <a
        className="navbar-brand"
        href="/"
        onClick={[Function]}
      >
        <img
          alt="logo navbar"
          className="img-fluid navbar-icon"
          src="/assets/svg/avatar.svg"
        />
      </a>
      <button
        className="navbar-toggler"
        onClick={[Function]}
        type="button"
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <article
        className="collapse navbar-collapse"
        id="navbarSupportedContent"
      >
        <div
          className="navbar-nav text-center "
        >
          <a
            className="nav-item nav-link text-white"
            href="/marvel"
            onClick={[Function]}
          >
            Marvel
          </a>
          <a
            className="nav-item nav-link text-white"
            href="/dc"
            onClick={[Function]}
          >
            DC
          </a>
          <a
            className="nav-item nav-link text-white"
            href="/search"
            onClick={[Function]}
          >
            Search
          </a>
        </div>
        <div
          className="navbar-nav ms-auto"
        >
          <ul
            className="navbar-nav"
          >
            <span
              className="nav-item nav-link text-white text-center"
            >
              Hola, 
              Leandro
              <small
                style={
                  Object {
                    "fontSize": "20px",
                  }
                }
              >
                 ðŸ‘‹ 
              </small>
            </span>
            <button
              className="nav-item btn btn-danger"
              onClick={[Function]}
            >
              Cerrar sesiÃ³n 
              <img
                alt="logout"
                src="/assets/svg/logout.svg"
              />
            </button>
          </ul>
        </div>
      </article>
    </section>
  </nav>,
  <main
    className="container p-4"
  >
    <div>
      <h1
        className="text-center"
      >
         Superheroes favoritos 
      </h1>
      <hr />
      <div
        className="container-card"
      >
        <div
          className="col animate__animated animate__fadeIn px-2"
        >
          <li
            className="card"
          >
            <div
              className="d-flex justify-content-end align-items-center"
            >
              <button
                className="main-container--button"
              >
                <img
                  alt="delete"
                  className="main-container--button-close"
                  onClick={[Function]}
                  src="/assets/svg/close.svg"
                />
              </button>
              <div
                className="Toastify"
              />
            </div>
            <img
              alt="Batman"
              className="card-img-top img-thumbnail"
              src="/assets/img/dc-batman.jpg"
            />
            <div
              className="card-body d-flex flex-column justify-content-center align-items-center"
            >
              <h5
                className="card-title"
              >
                 
                Batman
                 
              </h5>
              <p
                className="card-text"
              >
                 
                DC Comics
                 
              </p>
              <div
                className="d-flex flex-row "
              >
                <p
                  className="card-text p-2"
                >
                  <small
                    className="text-muted"
                  >
                     
                    Bruce Wayne
                     
                  </small>
                </p>
              </div>
              <p
                className="card-text p-2"
              >
                 
                Detective Comics #27
                 
              </p>
            </div>
          </li>
        </div>
      </div>
    </div>
  </main>,
]
`;

exports[`Testing on <PrivateRouter /> This test should show the component <DashboardRoute /> without any hero 1`] = `
Array [
  <nav
    className="navbar navbar-expand-lg navbar-dark bg-dark px-2"
  >
    <section
      className="container-fluid"
    >
      <a
        className="navbar-brand"
        href="/"
        onClick={[Function]}
      >
        <img
          alt="logo navbar"
          className="img-fluid navbar-icon"
          src="/assets/svg/avatar.svg"
        />
      </a>
      <button
        className="navbar-toggler"
        onClick={[Function]}
        type="button"
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <article
        className="collapse navbar-collapse"
        id="navbarSupportedContent"
      >
        <div
          className="navbar-nav text-center "
        >
          <a
            className="nav-item nav-link text-white"
            href="/marvel"
            onClick={[Function]}
          >
            Marvel
          </a>
          <a
            className="nav-item nav-link text-white"
            href="/dc"
            onClick={[Function]}
          >
            DC
          </a>
          <a
            className="nav-item nav-link text-white"
            href="/search"
            onClick={[Function]}
          >
            Search
          </a>
        </div>
        <div
          className="navbar-nav ms-auto"
        >
          <ul
            className="navbar-nav"
          >
            <span
              className="nav-item nav-link text-white text-center"
            >
              Hola, 
              Leandro
              <small
                style={
                  Object {
                    "fontSize": "20px",
                  }
                }
              >
                 ðŸ‘‹ 
              </small>
            </span>
            <button
              className="nav-item btn btn-danger"
              onClick={[Function]}
            >
              Cerrar sesiÃ³n 
              <img
                alt="logout"
                src="/assets/svg/logout.svg"
              />
            </button>
          </ul>
        </div>
      </article>
    </section>
  </nav>,
  <main
    className="container p-4"
  >
    <div
      className="d-flex justify-content-between align-items-center w-100 p-5 flex-column flex-lg-row "
    >
      <img
        alt="have not element"
        className="img-fluid"
        src="/assets/img/not-super-hero.png"
      />
      <p
        className="text-center fs-3 mt-5 mt-lg-0"
      >
        No hay superheroes en la lista de favoritos...
      </p>
    </div>
  </main>,
]
`;
